core/chains/evm/gas/models.go

```diff
@@ -71,19 +71,31 @@ func NewEstimator(lggr logger.Logger, ethClient evmclient.Client, cfg Config, ge
 	if rollups.IsRollupWithL1Support(cfg.ChainType()) {
 		l1Oracle = rollups.NewL1GasPriceOracle(lggr, ethClient, cfg.ChainType())
 	}
+	var est func(logger.Logger) EvmEstimator
 	switch s {
 	case "Arbitrum":
-		return NewWrappedEvmEstimator(NewArbitrumEstimator(lggr, geCfg, ethClient, ethClient), df, l1Oracle)
+		est = func(l logger.Logger) EvmEstimator {
+			return NewArbitrumEstimator(lggr, geCfg, ethClient, ethClient)
+		}
 	case "BlockHistory":
-		return NewWrappedEvmEstimator(NewBlockHistoryEstimator(lggr, ethClient, cfg, geCfg, bh, *ethClient.ConfiguredChainID()), df, l1Oracle)
+		est = func(l logger.Logger) EvmEstimator {
+			return NewBlockHistoryEstimator(lggr, ethClient, cfg, geCfg, bh, *ethClient.ConfiguredChainID())
+		}
 	case "FixedPrice":
-		return NewWrappedEvmEstimator(NewFixedPriceEstimator(geCfg, bh, lggr), df, l1Oracle)
+		est = func(l logger.Logger) EvmEstimator {
+			return NewFixedPriceEstimator(geCfg, bh, lggr)
+		}
 	case "L2Suggested", "SuggestedPrice":
-		return NewWrappedEvmEstimator(NewSuggestedPriceEstimator(lggr, ethClient), df, l1Oracle)
+		est = func(l logger.Logger) EvmEstimator {
+			return NewSuggestedPriceEstimator(lggr, ethClient)
+		}
 	default:
 		lggr.Warnf("GasEstimator: unrecognised mode '%s', falling back to FixedPriceEstimator", s)
-		return NewWrappedEvmEstimator(NewFixedPriceEstimator(geCfg, bh, lggr), df, l1Oracle)
+		est = func(l logger.Logger) EvmEstimator {
+			return NewFixedPriceEstimator(geCfg, bh, lggr)
+		}
 	}
+	return NewWrappedEvmEstimator(lggr, est, df, l1Oracle)
```

Comment:
What do you think of a factory pattern?

Interface:
```
type EstimatorFactory interface {
    CreateEstimator(logger.Logger, evmclient.Client, Config, evmconfig.GasEstimator, rollups.L1Oracle) EvmEstimator
}
```

Example Implementations
```
// ArbitrumEstimatorFactory creates an ArbitrumEstimator
type ArbitrumEstimatorFactory struct{}

func (f *ArbitrumEstimatorFactory) CreateEstimator(lggr logger.Logger, ethClient evmclient.Client, cfg Config, geCfg evmconfig.GasEstimator, l1Oracle rollups.L1Oracle) EvmEstimator {
    return NewArbitrumEstimator(lggr, geCfg, ethClient, ethClient)
}

// BlockHistoryEstimatorFactory creates a BlockHistoryEstimator
type BlockHistoryEstimatorFactory struct{}

func (f *BlockHistoryEstimatorFactory) CreateEstimator(lggr logger.Logger, ethClient evmclient.Client, cfg Config, geCfg evmconfig.GasEstimator, l1Oracle rollups.L1Oracle) EvmEstimator {
    bh := geCfg.BlockHistory()
    return NewBlockHistoryEstimator(lggr, ethClient, cfg, geCfg, bh, *ethClient.ConfiguredChainID())
}

// ... 
```
Switch case refactored: 
```
    var factory EstimatorFactory
    switch s {
    case "Arbitrum":
        factory = &ArbitrumEstimatorFactory{}
    case "BlockHistory":
        factory = &BlockHistoryEstimatorFactory{}
    // ... other cases
    default:
        lggr.Warnf("GasEstimator: unrecognised mode '%s', falling back to FixedPriceEstimator", s)
        factory = &FixedPriceEstimatorFactory{}
    }

    est := factory.CreateEstimator(lggr, ethClient, cfg, geCfg, l1Oracle)
    return NewWrappedEvmEstimator(lggr, est, df, l1Oracle)
```

Reply:
An interface seems like overkill since there is no state to track in the factories. But regardless, the crux of the problem here is the service name derived from the logger. The code was acting like the wrapper owned the estimator, but the service name did not reflect that. This change was meant to correct that by moving estimator construction inside of the wrapper. We could pass the factory in, but then we have to pipe along extra parameters through the wrapper.

Reply:
> We could pass the factory in, but then we have to pipe along extra parameters through the wrapper.

Could you give an example? I'm not seeing where the extra parameters would come from.

Reply:
Moved to #11558 and named better there.